app/src/main/java/com/pvp/app/ui/screen/goals/GoalViewModel.kt

```diff
@@ -0,0 +1,218 @@
+package com.pvp.app.ui.screen.goals
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.pvp.app.api.Configuration
+import com.pvp.app.api.GoalService
+import com.pvp.app.api.HealthConnectService
+import com.pvp.app.api.UserService
+import com.pvp.app.model.Goal
+import com.pvp.app.model.SportActivity
+import com.pvp.app.model.User
+import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.flatMapLatest
+import kotlinx.coroutines.flow.flowOf
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+import java.time.DayOfWeek
+import java.time.Instant
+import java.time.LocalDate
+import java.time.ZoneId
+import java.time.temporal.TemporalAdjusters
+import javax.inject.Inject
+
+@HiltViewModel
+@OptIn(ExperimentalCoroutinesApi::class)
+class GoalViewModel @Inject constructor(
+    configuration: Configuration,
+    private val goalService: GoalService,
+    private val healthConnectService: HealthConnectService,
+    private val userService: UserService
+) : ViewModel() {
+
+    private val _state = MutableStateFlow(GoalState())
+    val state = _state.asStateFlow()
+
+    val rangeKilometers = configuration.rangeKilometers
+
+    init {
+        viewModelScope.launch {
+            val flowUser = userService.user
+
+            val flowGoals = flowUser.flatMapLatest { user ->
+                user
+                    ?.let { goalService.get(user.email) }
+                    ?: flowOf(listOf())
+            }
+
+            flowUser
+                .combine(flowGoals) { user, goals ->
+                    user ?: return@combine _state.value
```

Comment:
Instead of such hacky return we could add `Flow#filterNotNull` onto the `userService.user` and then everything will be combined and collected only if user is not null. Then there will be no need to return empty goals list too.

Just reminding that `combine` waits until very first emit from any flow happens. So in case we know that user could possibly be null here for a long time, then thread would stay frozen, waiting. But I believe that here is not such case.

Also, could add `Dispatchers.IO` into `viewModelScope.launch(...)` so that everything happens off-main thread.

Reply:
Since `CalendarWeeklyViewModel` and `CalendarMonthlyViewModel` both use the same return, will modify them too.